source/Detector/Amptek/AmptekSDD123ServerGroup.cpp

```diff
@@ -0,0 +1,166 @@
+		AmptekSDD123ConfigurationMap *amptekConfiguration = configurationMaps.at(index);
+
+		amptekServer = new AmptekSDD123Server(amptekConfiguration, QString("%1 Server").arg(amptekConfiguration->detectorName()), this);
+		amptekServer->setRequestIntervalTimer(masterRequestInterval_);
+
+		servers_.append(amptekServer);
+		serversInDwellMode_.append(amptekServer);
+
+		serverAboutToConfigurationStateMapper_->setMapping(amptekServer, index);
+		serverConfigurationStateMapper_->setMapping(amptekServer, index);
+		serverDwellStateMapper_->setMapping(amptekServer, index);
+
+		connect(amptekServer, SIGNAL(serverAboutToChangeToConfigurationState()), serverAboutToConfigurationStateMapper_, SLOT(map()));
+		connect(amptekServer, SIGNAL(serverChangedToConfigurationState()), serverConfigurationStateMapper_, SLOT(map()));
+		connect(amptekServer, SIGNAL(serverChangedToDwellState()), serverDwellStateMapper_, SLOT(map()));
+	}
+
+	connect(serverAboutToConfigurationStateMapper_, SIGNAL(mapped(int)), this, SLOT(onServerAboutToChangeToConfigurationState(int)));
+	connect(serverConfigurationStateMapper_, SIGNAL(mapped(int)), this, SLOT(onServerChangedToConfigurationState(int)));
+	connect(serverDwellStateMapper_, SIGNAL(mapped(int)), this, SLOT(onServerChangedToDwellState(int)));
+
+	QTimer::singleShot(1000, this, SLOT(initiateAllRequestSpectrum()));
+}
+
+AmptekSDD123ServerGroup::~AmptekSDD123ServerGroup()
+{
+	releaseRequestSpectrualTimer();
+
+	foreach (AmptekSDD123Server *amptekServer, servers_) {
+		disconnect(amptekServer, SIGNAL(serverAboutToChangeToConfigurationState()), serverAboutToConfigurationStateMapper_, SLOT(map()));
+		disconnect(amptekServer, SIGNAL(serverChangedToConfigurationState()), serverConfigurationStateMapper_, SLOT(map()));
+		disconnect(amptekServer, SIGNAL(serverChangedToDwellState()), serverDwellStateMapper_, SLOT(map()));
+
+		amptekServer->deleteLater();
+	}
+	serversInDwellMode_.clear();
+	servers_.clear();
+
+
+	disconnect(serverAboutToConfigurationStateMapper_, SIGNAL(mapped(int)), this, SLOT(onServerAboutToChangeToConfigurationState(int)));
+	disconnect(serverConfigurationStateMapper_, SIGNAL(mapped(int)), this, SLOT(onServerChangedToConfigurationState(int)));
+	disconnect(serverDwellStateMapper_, SIGNAL(mapped(int)), this, SLOT(onServerChangedToDwellState(int)));
+
+	serverAboutToConfigurationStateMapper_->deleteLater();
+	serverConfigurationStateMapper_->deleteLater();
+	serverDwellStateMapper_->deleteLater();
+}
+
+AmptekSDD123Server* AmptekSDD123ServerGroup::serverAt(int index){
+	return servers_.at(index);
+}
+
+void AmptekSDD123ServerGroup::setServerState(int index, AmptekSDD123ServerGroup::ServerState serverState){
+	AmptekSDD123Server *amptekServer = serverAt(index);
+
+	if(serverState == AmptekSDD123ServerGroup::ConfigurationState){
+		if(serversInDwellMode_.contains(amptekServer))
+			serversInDwellMode_.removeAll(amptekServer);
+
+		QTimer::singleShot(masterRequestInterval_+5, amptekServer, SLOT(disableMCAMCS()));
```

Comment:
If I recall correctly, this was intentional. I think I discovered in testing that it was possible to put the server group in its ::ConfigurationState while it still wanted to send back the next packets of data. This meant that calling disableMCAMCS() too soon would leave junk in the buffer and the whole thing would fall apart. Making sure that this was called "a nominally short period of time after the latest time that new data could come back (ie,, masterRequestInterval)" meant that everything worked fine. To be clear, the [dis/en]ableMCAMCS() almost amounts to putting the Amptek into and out of continuous mode. You need to safely take it out of continuous mode to start sending configuration data to it.